# Organización de Datos (75.06/95.58)
## Primer Cuatrimestre 2021
### Trabajo Práctico N°1
#### Curso 1: Argerich

----

##### Introducción:
En el año 2015 Nepal fue afectado por el terremoto Gorkha, un sismo que registró una
magnitud de 7.8 en la escala Richter y tuvo su epicentro en la ciudad de Kathmandu.
Aproximadamente 600,000 estructuras en el centro y pueblos aledaños fueron dañadas o
destruidas. Un análisis posterior al sismo llevado por la Comisión Nacional de
Planeamiento de Nepal comunicó que la pérdida total económica ocasionada por el
terremoto fue de aproximadamente $7 mil millones (USD; NPC, 2015).
El dataset para el presente TP está compuesto de encuestas realizadas por Kathmandu
Living Labs y el Central Bureau of Statistics y contiene información sobre el impacto del
terremoto, estado de viviendas y estadísticas sociodemográficas.
Particularmente el dataset se enfoca en cómo eran las condiciones de una determinada vivienda y cuál fue su grado de daño luego del accidente. 

##### Objetivo:
El TP consiste en realizar un análisis exploratorio de los datos provistos con el objetivo de
determinar características y variables importantes, descubrir insights interesantes, y
analizar la estructura de los mismos.

# Análisis exploratorio de los datos
#### Empecemos importando los recursos necesarios

In [1]:
# importacion general de librerias y de visualizacion (matplotlib y seaborn)
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import matplotlib.ticker as mtick
import matplotlib as mpl
import cycler
import xgboost as xgb
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error
from sklearn import tree
from sklearn.model_selection import train_test_split

# for preprocessing the data
from sklearn.preprocessing import StandardScaler

# the model
from sklearn.ensemble import RandomForestClassifier

# for combining the preprocess with model training
from sklearn.pipeline import make_pipeline

# for optimizing the hyperparameters of the pipeline
from sklearn.model_selection import GridSearchCV

#seteando la paleta de colores
estilo = "mako"
cmap_cycler = sns.color_palette(estilo, n_colors = 280)
cmap = sns.color_palette(estilo, n_colors = 2000)[:500:-1]


plt.rcParams['axes.grid'] = True
mpl.rcParams['axes.prop_cycle'] = cycler.cycler('color', cmap_cycler[200::-20])
cmap_horizontal = mpl.rcParams['axes.prop_cycle'][::5]

In [2]:
pd.set_option('display.max_columns', None)

In [3]:
# lectura de archivos csv
train_values=pd.read_csv('train_values.csv',
                        dtype={
                            'geo_level_1_id':'int8',
                            'geo_level_2_id':'int16',
                            'geo_level_3_id':'int16',
                            'count_floors_pre_eq':'int8',
                            'age':'int16',
                            'area_percentage':'int8',
                            'height_percentage':'int8',
                            'land_surface_condition':'category',                 
                            'foundation_type':'category',                          
                            'roof_type':'category',                                 
                            'ground_floor_type':'category',                        
                            'other_floor_type':'category',                          
                            'position':'category',                                  
                            'plan_configuration':'category',
                            'has_superstructure_adobe_mud':'bool',               
                            'has_superstructure_mud_mortar_stone':'bool',        
                            'has_superstructure_stone_flag':'bool',             
                            'has_superstructure_cement_mortar_stone':'bool',     
                            'has_superstructure_mud_mortar_brick':'bool',        
                            'has_superstructure_cement_mortar_brick':'bool',     
                            'has_superstructure_timber':'bool',                  
                            'has_superstructure_bamboo':'bool',                  
                            'has_superstructure_rc_non_engineered':'bool',       
                            'has_superstructure_rc_engineered':'bool',          
                            'has_superstructure_other':'bool',
                            'legal_ownership_status':'category',
                            'count_families':'int8',
                            'has_secondary_use':'bool',                          
                            'has_secondary_use_agriculture':'bool',             
                            'has_secondary_use_hotel':'bool',                    
                            'has_secondary_use_rental':'bool',                  
                            'has_secondary_use_institution':'bool',             
                            'has_secondary_use_school':'bool',                
                            'has_secondary_use_industry':'bool',              
                            'has_secondary_use_health_post':'bool',         
                            'has_secondary_use_gov_office':'bool',           
                            'has_secondary_use_use_police':'bool',             
                            'has_secondary_use_other':'bool'                   
                        })
train_labels=pd.read_csv('train_labels.csv',
                        dtype={
                            'building_id':'int64',
                            'damage_grade':'int8'
                        })

### Limpieza de datos
- ¿Hay datos faltantes?
    - ¿Son muchos? Caso contrario podríamos simplemente filtrarlos.
- ¿Todos los datos están con el formato correcto?
    - ¿Es necesario transformar alguna columna a tipo categórico/numérico/date_time?
- ¿Los dataframes son muy grandes?
    - ¿Es necesario algun trabajo previo para minimizar el espacio en memoria?
- ¿Hay datos que no tengan sentido?
    - Cuyos valores sean muy grandes para el contexto
    - Que no sean posibles

#### Reviso valores null en mis dataframes

In [4]:
train_labels.isnull().any()

building_id     False
damage_grade    False
dtype: bool

In [5]:
train_values.isnull().any()

building_id                               False
geo_level_1_id                            False
geo_level_2_id                            False
geo_level_3_id                            False
count_floors_pre_eq                       False
age                                       False
area_percentage                           False
height_percentage                         False
land_surface_condition                    False
foundation_type                           False
roof_type                                 False
ground_floor_type                         False
other_floor_type                          False
position                                  False
plan_configuration                        False
has_superstructure_adobe_mud              False
has_superstructure_mud_mortar_stone       False
has_superstructure_stone_flag             False
has_superstructure_cement_mortar_stone    False
has_superstructure_mud_mortar_brick       False
has_superstructure_cement_mortar_brick  

Podemos ver que, por suerte, **ninguno de los dataframes contienen valores NaN.**  
Esto quiere decir que no será necesario ningún tipo de filtro en los datasets a causa de valores faltantes.

#### Chequeando formato de los datos

damage_grade originalmente era int64

In [6]:
train_labels.dtypes

building_id     int64
damage_grade     int8
dtype: object

Muchos valores categóricos eran objects y muchos valores booleanos eran int64, cambiamos sus tipos para optimizar.

In [7]:
train_values.dtypes

building_id                                  int64
geo_level_1_id                                int8
geo_level_2_id                               int16
geo_level_3_id                               int16
count_floors_pre_eq                           int8
age                                          int16
area_percentage                               int8
height_percentage                             int8
land_surface_condition                    category
foundation_type                           category
roof_type                                 category
ground_floor_type                         category
other_floor_type                          category
position                                  category
plan_configuration                        category
has_superstructure_adobe_mud                  bool
has_superstructure_mud_mortar_stone           bool
has_superstructure_stone_flag                 bool
has_superstructure_cement_mortar_stone        bool
has_superstructure_mud_mortar_b

#### Manejo de memoria del dataset
El valor inicialmente (sin castear las columnas al leer el csv) era de 3.97 Mb. Ahora se ve que decreció.

In [8]:
(train_labels.memory_usage()/(2**20)).sum()

2.2368783950805664

El valor inicialmente era de 77.54 Mb podemos observar que ahora la memoria se usa de forma mucho mas óptima.

In [9]:
(train_values.memory_usage()/(2**20)).sum()

12.179516792297363

#### Buscando datos fuera de lo común  
Buscamos si hay valores que no tengan sentido.

##### ¿Hay algún building_id en un dataset que **no** esté en el otro?
Los comparamos y chequeamos si hay algún false, los sumamos.

In [10]:
(train_values['building_id'] != train_labels['building_id']).sum()

0

Observamos que hay correspondencia 1 a 1 entre ambos datasets.
Eso significa que se puede hacer un merge (lo haremos mas adelante)

##### ¿Las variables geo_level_id tiene los valores que dice el enunciado?

In [11]:
train_values[['geo_level_1_id','geo_level_2_id','geo_level_3_id']].max()

geo_level_1_id       30
geo_level_2_id     1427
geo_level_3_id    12567
dtype: int16

Vemos que se corresponde con el enunciado.

In [12]:
train_values[['geo_level_1_id','geo_level_2_id','geo_level_3_id']].min()

geo_level_1_id    0
geo_level_2_id    0
geo_level_3_id    0
dtype: int16

Se corresponde con el enunciado.

##### ¿Hay algún valor de cantidad de familias por hogar que nos llame la atención?
No nos llama la atencion que existan algunas edificaciones con mas de 7 familias, podrían ser por ejemplo bloques de departamentos.
Tambien observamos que hay 20 mil edificaciones sin familias, esto puede ser tranquilamente por ser edificaciones destinadas a otras actividades.

In [13]:
train_values['count_families'].value_counts()

1    226115
0     20862
2     11294
3      1802
4       389
5       104
6        22
7         7
9         4
8         2
Name: count_families, dtype: int64

##### ¿Hay valores raros en la cantidad de pisos por edificio?
No vemos que sea inverosimil que un solo edificio tenga 9 u 8 pisos.

In [14]:
train_values['count_floors_pre_eq'].value_counts()

2    156623
3     55617
1     40441
4      5424
5      2246
6       209
7        39
8         1
9         1
Name: count_floors_pre_eq, dtype: int64

##### ¿Area_percentage y Height_percentage están dentro de los valores [0,100]?
Como se observa a continuación, sí.

In [15]:
train_values[['area_percentage','height_percentage']].max()

area_percentage      100
height_percentage     32
dtype: int8

In [16]:
train_values[['area_percentage','height_percentage']].min()

area_percentage      1
height_percentage    2
dtype: int8

##### ¿Hay valores fuera de lo normal para la edad de las edificaciones?

In [17]:
train_values['age'].value_counts()

10     38896
15     36010
5      33697
20     32182
0      26041
25     24366
30     18028
35     10710
40     10559
50      7257
45      4711
60      3612
80      3055
55      2033
70      1975
995     1390
100     1364
65      1123
90      1085
85       847
75       512
95       414
120      180
150      142
200      106
110      100
105       89
125       37
115       21
130        9
140        9
180        7
170        6
160        6
175        5
135        5
190        3
145        3
195        2
165        2
185        1
155        1
Name: age, dtype: int64

Si bien podemos observar que hay muchos edificios con antigüedades entre 0 y 200 años (lo cual
consideramos normal) podemos ver que **hay 1390 con antigüedades** de 995 años.  
Lo que nos llama la atención de esto es la brecha entre el valor 200 y 995.
- ¿Por qué no hay valores intermedios? 
- ¿Es posible que una edificación sea tan vieja?  

La forma en la que decidimos interpretar estos datos es que no se tiene una fecha concreta de la construcción de estas viviendas, pero que al ser tan antiguas (más de 200 años) se decidió ingresarlas con edades de 995 años (por eso tantas tienen exactamente esa edad).  
Para el resto de nuestro análisis consideraremos a esas viviendas como simplemente *viviendas muy viejas*

In [18]:
train_values.head()

building_id  geo_level_1_id  geo_level_2_id  geo_level_3_id  \
0       802906               6             487           12198   
1        28830               8             900            2812   
2        94947              21             363            8973   
3       590882              22             418           10694   
4       201944              11             131            1488   

   count_floors_pre_eq  age  area_percentage  height_percentage  \
0                    2   30                6                  5   
1                    2   10                8                  7   
2                    2   10                5                  5   
3                    2   10                6                  5   
4                    3   30                8                  9   

  land_surface_condition foundation_type roof_type ground_floor_type  \
0                      t               r         n                 f   
1                      o               r         n                 x   
2                      t               r         n                 f   
3                      t               r         n                 f   
4                      t               r         n                 f   

  other_floor_type position plan_configuration  has_superstructure_adobe_mud  \
0                q        t                  d                          True   
1                q        s                  d                         False   
2                x        t                  d                         False   
3                x        s                  d                         False   
4                x        s                  d                          True   

   has_superstructure_mud_mortar_stone  has_superstructure_stone_flag  \
0                                 True                          False   
1                                 True                          False   
2                                 True                          False   
3                                 True                          False   
4                                False                          False   

   has_superstructure_cement_mortar_stone  \
0                                   False   
1                                   False   
2                                   False   
3                                   False   
4                                   False   

   has_superstructure_mud_mortar_brick  \
0                                False   
1                                False   
2                                False   
3                                False   
4                                False   

   has_superstructure_cement_mortar_brick  has_superstructure_timber  \
0                                   False                      False   
1                                   False                      False   
2                                   False                      False   
3                                   False                       True   
4                                   False                      False   

   has_superstructure_bamboo  has_superstructure_rc_non_engineered  \
0                      False                                 False   
1                      False                                 False   
2                      False                                 False   
3                       True                                 False   
4                      False                                 False   

   has_superstructure_rc_engineered  has_superstructure_other  \
0                             False                     False   
1                             False                     False   
2                             False                     False   
3                             False                     False   
4                             False                     False   

  legal_ownership_status  count_families  has_secondary_use  \
0                 

In [19]:
train_labels.head()

building_id  damage_grade
0       802906             3
1        28830             2
2        94947             3
3       590882             2
4       201944             3

#### Mergeamos los dos dataframes ya que tiene una correspondencia 1 a 1

In [20]:
df=train_values.merge(train_labels)
df.head()

building_id  geo_level_1_id  geo_level_2_id  geo_level_3_id  \
0       802906               6             487           12198   
1        28830               8             900            2812   
2        94947              21             363            8973   
3       590882              22             418           10694   
4       201944              11             131            1488   

   count_floors_pre_eq  age  area_percentage  height_percentage  \
0                    2   30                6                  5   
1                    2   10                8                  7   
2                    2   10                5                  5   
3                    2   10                6                  5   
4                    3   30                8                  9   

  land_surface_condition foundation_type roof_type ground_floor_type  \
0                      t               r         n                 f   
1                      o               r         n                 x   
2                      t               r         n                 f   
3                      t               r         n                 f   
4                      t               r         n                 f   

  other_floor_type position plan_configuration  has_superstructure_adobe_mud  \
0                q        t                  d                          True   
1                q        s                  d                         False   
2                x        t                  d                         False   
3                x        s                  d                         False   
4                x        s                  d                          True   

   has_superstructure_mud_mortar_stone  has_superstructure_stone_flag  \
0                                 True                          False   
1                                 True                          False   
2                                 True                          False   
3                                 True                          False   
4                                False                          False   

   has_superstructure_cement_mortar_stone  \
0                                   False   
1                                   False   
2                                   False   
3                                   False   
4                                   False   

   has_superstructure_mud_mortar_brick  \
0                                False   
1                                False   
2                                False   
3                                False   
4                                False   

   has_superstructure_cement_mortar_brick  has_superstructure_timber  \
0                                   False                      False   
1                                   False                      False   
2                                   False                      False   
3                                   False                       True   
4                                   False                      False   

   has_superstructure_bamboo  has_superstructure_rc_non_engineered  \
0                      False                                 False   
1                      False                                 False   
2                      False                                 False   
3                       True                                 False   
4                      False                                 False   

   has_superstructure_rc_engineered  has_superstructure_other  \
0                             False                     False   
1                             False                     False   
2                             False                     False   
3                             False                     False   
4                             False                     False   

  legal_ownership_status  count_families  has_secondary_use  \
0                 

# RandomForest
#### Híper parámetros
* n_estimators: cantidad de árboles a construir
* max_depth: máxima profundidad de cada árbol
* min_samples_split: la cantidad mínima de datos requeridos para splitear un nodo interno
* min_samples_leaf: cantidad mínima de datos requeridos para ser una hoja
* max_features: la cantidad de features a considerar cuando se busca el mejor split

## Manipulamos los features categóricos

In [21]:
one_hotear = {'geo_level_1_id': 'geo1', 'land_surface_condition': 'lsc',
             'roof_type':'roof', 'foundation_type': 'foundation',
             'ground_floor_type':'ground_floor',
             'other_floor_type':'other_floor', 'position':'pos',
             'legal_ownership_status':'ownership',
             'plan_configuration':'plan'}

In [22]:
a_dropear = ['geo_level_2_id', 'geo_level_3_id', 'building_id']

In [23]:
def one_hot(df, feature, prefix):
    y = pd.get_dummies(df[feature], prefix=prefix)
    df = df.drop(feature, axis = 1)
    df = df.join(y)
    return df

In [24]:
def encodear(df, d):
    for key,value in d.items():
        df = one_hot(df, key, value)
    return df

In [25]:
def dropear(df, l):
    for feature in l:
        df = df.drop(feature, axis=1)
    return df

In [26]:
df = encodear(df, one_hotear)
df = dropear(df, a_dropear)

In [27]:
df.head()

count_floors_pre_eq  age  area_percentage  height_percentage  \
0                    2   30                6                  5   
1                    2   10                8                  7   
2                    2   10                5                  5   
3                    2   10                6                  5   
4                    3   30                8                  9   

   has_superstructure_adobe_mud  has_superstructure_mud_mortar_stone  \
0                          True                                 True   
1                         False                                 True   
2                         False                                 True   
3                         False                                 True   
4                          True                                False   

   has_superstructure_stone_flag  has_superstructure_cement_mortar_stone  \
0                          False                                   False   
1                          False                                   False   
2                          False                                   False   
3                          False                                   False   
4                          False                                   False   

   has_superstructure_mud_mortar_brick  \
0                                False   
1                                False   
2                                False   
3                                False   
4                                False   

   has_superstructure_cement_mortar_brick  has_superstructure_timber  \
0                                   False                      False   
1                                   False                      False   
2                                   False                      False   
3                                   False                       True   
4                                   False                      False   

   has_superstructure_bamboo  has_superstructure_rc_non_engineered  \
0                      False                                 False   
1                      False                                 False   
2                      False                                 False   
3                       True                                 False   
4                      False                                 False   

   has_superstructure_rc_engineered  has_superstructure_other  count_families  \
0                             False                     False               1   
1                             False                     False               1   
2                             False                     False               1   
3                             False                     False               1   
4                             False                     False               1   

   has_secondary_use  has_secondary_use_agriculture  has_secondary_use_hotel  \
0              False                          False                    False   
1              False                          False                    False   
2              False                          False                    False   
3              False                          False                    False   
4              False                          False                    False   

   has_secondary_use_rental  has_secondary_use_institution  \
0                     False                          False   
1                     False                          False   
2                     False                          False   
3                     False                          False   
4                     False                          False   

   has_secondary_use_school  has_secondary_use_industry  \
0                     False                       False   
1                     False                       False   
2                     False                       False   
3                     Fals

In [28]:
df = df.drop('damage_grade', axis=1)

In [29]:
data = df
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 260601 entries, 0 to 260600
Data columns (total 96 columns):
 #   Column                                  Non-Null Count   Dtype
---  ------                                  --------------   -----
 0   count_floors_pre_eq                     260601 non-null  int8 
 1   age                                     260601 non-null  int16
 2   area_percentage                         260601 non-null  int8 
 3   height_percentage                       260601 non-null  int8 
 4   has_superstructure_adobe_mud            260601 non-null  bool 
 5   has_superstructure_mud_mortar_stone     260601 non-null  bool 
 6   has_superstructure_stone_flag           260601 non-null  bool 
 7   has_superstructure_cement_mortar_stone  260601 non-null  bool 
 8   has_superstructure_mud_mortar_brick     260601 non-null  bool 
 9   has_superstructure_cement_mortar_brick  260601 non-null  bool 
 10  has_superstructure_timber               260601 non-null  bool 
 11  

In [30]:
data.head()
data.shape

(260601, 96)

In [31]:
from sklearn.metrics import f1_score

In [32]:
import pickle
filename = 'random_forest_1_pickle.sav'

In [33]:
pipe = make_pipeline(StandardScaler(), 
                     RandomForestClassifier(random_state=2018))
pipe

Pipeline(steps=[('standardscaler', StandardScaler()),
                ('randomforestclassifier',
                 RandomForestClassifier(random_state=2018))])

In [34]:
param_grid = {'randomforestclassifier__n_estimators': [50, 100],
              'randomforestclassifier__min_samples_leaf': [1, 5]}
gs = GridSearchCV(pipe, param_grid, cv=5)

In [35]:
train_labels = train_labels.drop('building_id', axis=1)

In [36]:
#gs.fit(data, train_labels.values.ravel())
#pickle.dump(gs, open(filename, 'wb'))

In [37]:
gs = pickle.load(open(filename, 'rb'))

In [38]:
gs.best_params_

{'randomforestclassifier__min_samples_leaf': 5,
 'randomforestclassifier__n_estimators': 100}

In [39]:
train_values.head()

building_id  geo_level_1_id  geo_level_2_id  geo_level_3_id  \
0       802906               6             487           12198   
1        28830               8             900            2812   
2        94947              21             363            8973   
3       590882              22             418           10694   
4       201944              11             131            1488   

   count_floors_pre_eq  age  area_percentage  height_percentage  \
0                    2   30                6                  5   
1                    2   10                8                  7   
2                    2   10                5                  5   
3                    2   10                6                  5   
4                    3   30                8                  9   

  land_surface_condition foundation_type roof_type ground_floor_type  \
0                      t               r         n                 f   
1                      o               r         n                 x   
2                      t               r         n                 f   
3                      t               r         n                 f   
4                      t               r         n                 f   

  other_floor_type position plan_configuration  has_superstructure_adobe_mud  \
0                q        t                  d                          True   
1                q        s                  d                         False   
2                x        t                  d                         False   
3                x        s                  d                         False   
4                x        s                  d                          True   

   has_superstructure_mud_mortar_stone  has_superstructure_stone_flag  \
0                                 True                          False   
1                                 True                          False   
2                                 True                          False   
3                                 True                          False   
4                                False                          False   

   has_superstructure_cement_mortar_stone  \
0                                   False   
1                                   False   
2                                   False   
3                                   False   
4                                   False   

   has_superstructure_mud_mortar_brick  \
0                                False   
1                                False   
2                                False   
3                                False   
4                                False   

   has_superstructure_cement_mortar_brick  has_superstructure_timber  \
0                                   False                      False   
1                                   False                      False   
2                                   False                      False   
3                                   False                       True   
4                                   False                      False   

   has_superstructure_bamboo  has_superstructure_rc_non_engineered  \
0                      False                                 False   
1                      False                                 False   
2                      False                                 False   
3                       True                                 False   
4                      False                                 False   

   has_superstructure_rc_engineered  has_superstructure_other  \
0                             False                     False   
1                             False                     False   
2                             False                     False   
3                             False                     False   
4                             False                     False   

  legal_ownership_status  count_families  has_secondary_use  \
0                 

In [40]:
in_sample_preds = gs.predict(data)
f1_score(train_labels, in_sample_preds, average='micro')

0.7204423620784265

## Probamos con el set de test

In [41]:
test_values = pd.read_csv('test_values.csv')

In [42]:
test_values = encodear(test_values, one_hotear)

In [43]:
test_values = dropear(test_values, a_dropear)

In [44]:
predictions = gs.predict(test_values)

In [45]:
predictions.size

86868

In [46]:
submission_format = pd.read_csv('submission_format.csv', index_col='building_id')

In [47]:
submission_format.size

86868

In [48]:
my_submission = pd.DataFrame(data=predictions,columns=submission_format.columns,index=submission_format.index)

In [49]:
my_submission

damage_grade
building_id              
300051                  3
99355                   2
890251                  2
745817                  1
421793                  3
...                   ...
310028                  2
663567                  2
1049160                 2
442785                  2
501372                  2

[86868 rows x 1 columns]

In [50]:
my_submission.to_csv('submission_1_random_forest.csv')

In [51]:
!head submission_1_random_forest.csv

building_id,damage_grade
300051,3
99355,2
890251,2
745817,1
421793,3
871976,2
691228,1
896100,3
343471,2
